034-kafka-topic-replication-factor.md

```diff
@@ -0,0 +1,76 @@
+
+For instance, users can lower the RFs for selective less critical topics to reduce the infrastructure cost (Compute, Storage and Network). Also, for some critical kafka topics, users can increase the data reliability through the bump of the RFs ONLY for these specific topics. 
+
+More importantly, users only need to specify the new RF number in the `KafkaTopic` resource and approve the change, in order to change the RF for kafka topics. The heavy lifting work, such as new replica placement, extra replica deletion and possible cluster rebalance, is completely taken care by the strimzi. 
+
+## Proposal
+
+### API
+The Kafka Topic RF can be changed by users in the following steps:
+* Update the `spec.replicas` properties in the target `KafkaTopic` Resource in the kubernetes cluster
+* Watch the `Status` field of the `KafkaTopic Resource`, for the status update of this RF change.
+
+### Design
+We propose to change the topic RFs with the support of `Cruise Control` and `Strimzi Topic Operator`. 
+
+#### Cruise Control
+Cruise Control Service will be leveraged to change the RF for topics in the kafka clusters, for the following reasons: 
+* Strimzi supports the Cruise Control Service, and it's ready for use with simple configurations. 
+* Cruise Control has native support for the RF changes. It automates the whole process, including selecting of new replicas and extra replicas, triggering the replica creation in the kafka brokers and monitoring the cluster during the process.
+
+#### Kafka Topic Operator
+Strimzi Topic Operator will watch the changes of `KafkaTopic` Resource. 
+
+For any new topic RF change event, the `Strimzi Topic Operator` will validate the change. For instance, the new RF should be >= the min insync replica set size for the target topic. 
+
+For the valid changes, the `Strimzi Topic Operator` will send requests to `Cruise Control`. The Kafka Topic Operator will periodically poll the status from the Cruise Control, and update the status in the `Status` field of the `KafkaTopic` resource.
+
+For the invalid changes, the `Strimzi Topic Operator` will reject it and send the reason in the `Status` field of the `KafkaTopic` resources.
+
+### Work flow
+In the high level, the RF change will be performed in the following steps:
+* Users modify the value of the `spec.replicas` properties in the target `KafkaTopic` resource.
+* Kafka Topic Operator watches this resource changes and send the request to the Cruise Control Service.
+* Kafka Topic Operator will periodically poll the result from the Cruise Control Service, and update the status in the `KafkaTopic` resource.
+* Once it's completed by the Cruise Control Service, the Kafka Topic Operator will mark the Status in the `KafkaTopic` resource as `Completed`.
+
+## Affected/not affected projects
+
+The following components will be affected by this proposal:
+* Strimzi Topic Operator
+* Kafka Rebalance Operator
```

Comment:
From the proposal is not clear how the Kafka Rebalance Operator is affected. Are we assuming to factor out some common code to use in the TO?

Reply:
I think we need to look at what common code can be shared here as a lot of the interactions will be the same.

Reply:
Yes, it's correct.